content/blog/2025/07-01_httr2_client/index.qmd

```diff
@@ -0,0 +1,623 @@
+Which confused me, since I thought given I recorded the responses with vcr, that wouldn't happen?
+
+Where I got my logic wrong, was that what was failing wasn't the API calls (or vcr's accessing of the cassettes), but rather my own checking function for whether authentication was set up (which it was not on GitHub Actions!)
+
+Remember I made a `ns_has_auth` function, that checks if authentication is set up?
+Well, that's integrated in the `ns_client` function, meaning _before_ any calls to the API!
+So my GitHub Actions. were failing before any cassettes were in play.
+
+This is where the dreaded moment I knew was going to happen in this journey would indeed happen, I needed to figure out how to _mock_ a function.
+Mocking in a test is used to circumvent a specific piece of code from testing.
+This sounds counterintuitive, since we are supposed to be testing the code, but it is very necessary in certain cases.
+
+In this particular case, I don't need to test the `ns_has_auth()` function, that is not the functionality I am after testing.
+I can test that in another test specifically made to test it, where I can control input and results better.
+In this case, I want to test the API call and make sure vcr has recorded a good cassette and that this cassette is used in instances where I need it to (like on GitHub Actions. and during CRAN checks).
+
+With all that being said, I need to make sure that the `ns_has_auth()` function always returns true and can continue on to the remaining function where I ask it to during testing.
+We create a function with `testthat::local_mocked_bindings` which we will wrap around any code where we want to test the other parts of the function rather than checking if authentication is set up.
+The functions defined in the locked mocked bindings, will replace the function of the same name from the `.package` we specify.
+This makes it possible to _mock_ any function from any package when needed.
+
+```r
+with_mocked_nettskjema_auth <- function(expr) {
+  testthat::local_mocked_bindings(
+    ns_has_auth = function(...) TRUE,
+    .package = "nettskjemar"
+  )
+  force(expr)
+}
+```
+
+This function uses the base R `force` function to run the code it's wrapped around, using the `ns_has_auth()` function that has been set up as a local mocked binding.
+So anything wrapped in this code that calls `ns_has_auth()` will always return `TRUE` because I have mocked it that way.
+
+Our final test code, therefore, became a layered structure: a testthat block wrapping a vcr cassette, which in turn enveloped our mocked function call.
+
+
+```r
+test_that("test user information", {
+  vcr::use_cassette("ns_get_me", {
+    with_mocked_nettskjema_auth(
+      me <- ns_get_me()
+    )
+  })
+
+  expect_is(me, "list")
+  expect_length(me, 10)
+})
+```
+
+Now, the `ns_get_me` function I am trying to test, will always get to the API call part of the function, because I am circumventing the `ns_has_auth` function.
```

Comment:
```suggestion
Now, the `ns_get_me()` function I am trying to test, will always get to the API call part of the function, because I am circumventing the `ns_has_auth()` function.
```